options/customizable.cc

```diff
@@ -0,0 +1,139 @@
+// Copyright (c) 2011-present, Facebook, Inc. All rights reserved.
+//  This source code is licensed under both the GPLv2 (found in the
+//  COPYING file in the root directory) and Apache 2.0 License
+//  (found in the LICENSE.Apache file in the root directory).
+
+#include "rocksdb/customizable.h"
+
+#include "rocksdb/convenience.h"
+#include "rocksdb/status.h"
+#include "util/string_util.h"
+
+namespace ROCKSDB_NAMESPACE {
+
+std::string Customizable::GetOptionName(const std::string& long_name) const {
+  const std::string& name = Name();
+  size_t name_len = name.size();
+  if (long_name.size() > name_len + 1 &&
+      long_name.compare(0, name_len, name) == 0 &&
+      long_name.at(name_len) == '.') {
+    return long_name.substr(name_len + 1);
+  } else {
+    return Configurable::GetOptionName(long_name);
+  }
+}
+
+#ifndef ROCKSDB_LITE
+Status Customizable::GetOption(const ConfigOptions& config_options,
+                               const std::string& opt_name,
+                               std::string* value) const {
+  if (opt_name == kIdPropName) {
+    *value = GetId();
+    return Status::OK();
+  } else {
+    return Configurable::GetOption(config_options, opt_name, value);
+  }
+}
+
+std::string Customizable::SerializeOptions(const ConfigOptions& config_options,
+                                           const std::string& prefix) const {
+  std::string result;
+  std::string parent;
+  if (!config_options.IsShallow()) {
+    parent = Configurable::SerializeOptions(config_options, "");
+  }
+  if (parent.empty()) {
+    result = GetId();
```

Comment:
No prefix here?

Reply:
No prefix is correct.  If it is shallow, we are not serializing the options contained in the object and only need to serialize the ID of the object and not the object in its entirety.

In a future PR, I will introduce more ways to serialize the objects.  At the moment, most complex objects are serialized in a "struct" form (  property = { field1 = value1; field2 = value2}.  I would like to be able to serialize things "detached":
   property.field1 = value1;
   property.field2 = value2;
and additionally allow prefixes on the objects:
  table.block_based.opt1 = value1;
  table.plain.opt1 = value2;

I would also like to be able to serialize options in "flags" format for command line options (--field1=value1) or in "Dump" format for log files:
  property.field1        : value1

All of those formats and complexity to the processing of Serialize and were accounted for in a previous iteration.  I removed them for now to simplify the review and processing.